src/troute-bmi/src/troute_nwm_bmi/troute_bmi.py

```diff
@@ -0,0 +1,326 @@
+"""Basic Model Interface implementation for NGEN t-route."""
+import numpy as np
+from bmipy import Bmi
+
+from .troute_model import Model
+
+_COUNT_SUFFIX = "__count"
+
+_VAR_NAME_UNITS_MAP = {
+    'land_surface_water_source__volume_flow_rate': ['streamflow_cms', 'm3 s-1'],
+}
+
+class BmiTroute(Bmi):
+    _model: Model
+
+    def __init__(self):
+        super().__init__()
+        self._values = {
+            "land_surface_water_source__id": np.zeros(0, dtype=np.intc),
+            "land_surface_water_source__volume_flow_rate": np.zeros(0, dtype=float),
+            "upstream_id": np.zeros(0, dtype=int),
+        }
+        self._var_loc = "node"
+        self._var_grid_id = 0
+        self._time_units = "s"
+        self._start_time = 0.0
+        self._end_time = np.finfo("d").max
+
+
+    def initialize(self, bmi_cfg_file):
+        self._model = Model(bmi_cfg_file)
+        self._var_units_map = {
+            name: units[1] for name, units in _VAR_NAME_UNITS_MAP.items()
+        }
+
+    def update(self):
+        self._model.update(self._values)
+    
+    def update_until(self, time):
+        while self._time < time:
+            self.update()
+
+    def set_value(self, var_name: str, src):
+        """
+        Set model values
+        
+        Parameters
+        ----------
+        var_name : str
+            Name of variable as CSDMS Standard Name.
+        src : array_like
+            Array of new values.
+        """
+        # special case for changing data size 
+        if var_name.endswith(_COUNT_SUFFIX):
+            source_var_name = var_name[:-len(_COUNT_SUFFIX)]
+            source = self._values.get(source_var_name)
+            self._values[source_var_name] = np.resize(source, int(src[0]))
+        else:
+            self._values[var_name] = src
+
+    def get_value(self, var_name):
+        """Copy of values.
+        Parameters
+        ----------
+        var_name : str
+            Name of variable as CSDMS Standard Name.
+        Returns
+        -------
+        output_df : pd.DataFrame
+            Copy of values.
+        """
+        output_df = self.get_value_ptr(var_name)
+        return output_df
+
+    def get_value_ptr(self, var_name: str):
+        """Reference to values.
+        Parameters
+        ----------
+        var_name : str
+            Name of variable as CSDMS Standard Name.
+        Returns
+        -------
+        array_like
+            Value array.
+        """
+        if var_name.endswith(_COUNT_SUFFIX):
+            source_var_name = var_name[:-len(_COUNT_SUFFIX)]
+            source = self._values[source_var_name]
+            return np.array([len(source)], dtype=np.int64)
+        else:
+            return self._values[var_name]
```

Comment:
If you're making more updates, this can just be 
```suggestion
        return self._values[var_name]
```
If not, no big deal

Reply:
Fill out more of the BMI functions. `get_value` will now always return a copy of the data whilst `get_value_ptr` will return the value itself. This makes the `__count` variant of the variables incompatible with `get_value_ptr`, which makes more sense for how the BMI is supposed to work.